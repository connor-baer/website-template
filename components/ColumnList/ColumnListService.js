import { get } from 'lodash/fp';

const GRID_COLUMNS = 12;

export function getNarrowTemplateAreas(columns) {
  return columns
    .map((_, index) => {
      const areaName = `column-${index} `;
      return `"${areaName.repeat(GRID_COLUMNS)}"`;
    })
    .join(' ');
}

export function getWideTemplateAreas(columns) {
  let remainingColumns = GRID_COLUMNS;

  const templateAreas = columns.reduce((areas, column, index) => {
    const areaName = `column-${index} `;
    const ratio = get('data.ratio', column);
    const isLast = index === columns.length - 1;
    const columnWidth = isLast
      ? remainingColumns
      : Math.round(GRID_COLUMNS * ratio);

    remainingColumns -= columnWidth;

    return `${areas}${areaName.repeat(columnWidth)}`;
  }, '');

  return `"${templateAreas}"`;
}
