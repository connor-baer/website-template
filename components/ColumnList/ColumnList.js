import styled from '@emotion/styled';
import { css } from '@emotion/core';

import * as ColumnListService from './ColumnListService';

const baseStyles = ({ theme, columns = [] }) => {
  const narrowTemplateAreas = ColumnListService.getNarrowTemplateAreas(columns);
  const wideTemplateAreas = ColumnListService.getWideTemplateAreas(columns);

  return css`
    grid-column: 1 / 13 !important;
    display: grid;
    grid-template-columns: repeat(12, [col-start] minmax(0, 1fr));
    grid-template-areas: ${narrowTemplateAreas};
    margin: 2rem 0;

    ${theme.mq.kilo} {
      grid-column-gap: 3rem;
      grid-template-areas: ${wideTemplateAreas};
    }
  `;
};

const ColumnList = styled('div')(baseStyles);

export default ColumnList;
