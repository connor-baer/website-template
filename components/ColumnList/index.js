import ColumnList from './ColumnList';

export default ColumnList;
