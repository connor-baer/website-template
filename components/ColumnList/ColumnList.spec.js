import React from 'react';

import ColumnList from './ColumnList';

describe('ColumnList', () => {
  function renderColumnList(fn, props = {}) {
    return fn(<ColumnList {...props} />);
  }

  it('should render with default styles', () => {
    const columns = [
      { data: { ratio: 0.5 } },
      { data: { ratio: 0.25 } },
      { data: { ratio: 0.25 } }
    ];
    const { container } = renderColumnList(render, { columns });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should meet basic accessibility guidelines', async () => {
    const columns = [
      { data: { ratio: 0.5 } },
      { data: { ratio: 0.25 } },
      { data: { ratio: 0.25 } }
    ];
    const html = renderColumnList(renderToHtml, { columns });
    const result = await axe(html);
    expect(result).toHaveNoViolations();
  });
});
