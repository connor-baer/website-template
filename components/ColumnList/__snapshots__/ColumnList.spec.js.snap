// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnList should render with default styles 1`] = `
.circuit-0 {
  grid-column: 1 / 13 !important;
  display: grid;
  grid-template-columns: repeat(12,[col-start] minmax(0,1fr));
  grid-template-areas: "column-0 column-0 column-0 column-0 column-0 column-0 column-0 column-0 column-0 column-0 column-0 column-0 " "column-1 column-1 column-1 column-1 column-1 column-1 column-1 column-1 column-1 column-1 column-1 column-1 " "column-2 column-2 column-2 column-2 column-2 column-2 column-2 column-2 column-2 column-2 column-2 column-2 ";
  margin: 2rem 0;
}

@media (min-width:600px) {
  .circuit-0 {
    grid-column-gap: 3rem;
    grid-template-areas: "column-0 column-0 column-0 column-0 column-0 column-0 column-1 column-1 column-1 column-2 column-2 column-2 ";
  }
}

<div
  class="circuit-0 circuit-1"
/>
`;
